### PR TITLE
Gpu extensions

### PR DESCRIPTION
**Pull request by julian-spahl**
# Description

Makes it possible to run Interactions on the GPU  
Added Data Layout for GPU  
Added Cuda Version of Lennard-Jones (12-6)  
Added specialized LinkedCells Traversal  
Updated CMake to optionally compile with Cuda support  

# How Has This Been Tested?

Please describe the tests that you ran to verify your changes.

- [x] All tests passed when compiled without Cuda Support  
- [x] Test for Data Transfer CPU - GPU
- [x] Test for Cuda LJ Functor for N3 and NoN3
- [x] Generic Tests Cuda Version vs. Non Cuda Versions

### CODE REVIEW

File: src/autopas/cells/FullParticleCell.h [hunk 1]
```diff
@@ -86,6 +87,16 @@ class FullParticleCell : public ParticleCell<Particle> {
    */
   SoA<SoAArraysType> _particleSoABuffer;
 
+  /**
```
Comment (SteffenSeckler) [inline src/autopas/cells/FullParticleCell.h]: should be present only if cuda is enabled

File: src/autopas/cells/FullParticleCell.h [hunk 2]
```diff
@@ -86,6 +87,16 @@ class FullParticleCell : public ParticleCell<Particle> {
...
+   * device AoS Buffer
+   */
+  double* _particlesDevice;
+
+  /**
+   * device particle SoABuffer
+   */
+  CudaSoA<typename Particle::CudaDeviceArraysType> _particleSoABufferDevice;
```
Comment (SteffenSeckler) [inline src/autopas/cells/FullParticleCell.h]: should also only be present if cuda is enabled

File: src/autopas/containers/cellPairTraversals/DummyTraversal.h [hunk 3]
```diff
@@ -31,6 +31,8 @@ class DummyTraversal : public CellPairTraversal<ParticleCell> {
   TraversalOption getTraversalType() override { return TraversalOption::dummyTraversal; }
 
   bool isApplicable() override { return true; }
+
+  DataLayoutOption requiredDataLayout() override { return DataLayoutOption::aos; }
```
Comment (SteffenSeckler) [inline src/autopas/containers/cellPairTraversals/DummyTraversal.h]: should not be needed, instead use isApplicable for this

File: src/autopas/containers/directSum/DirectSum.h [hunk 4]
```diff
@@ -107,6 +109,77 @@ class DirectSum : public ParticleContainer<Particle, ParticleCell> {
...
+  /**
+   * Function to iterate over all pairs of particles in an structure of arrays setting on the gpu
+   * @tparam ParticleFunctor
+   * @tparam Traversal
+   * @param f functor that describes the pair-potential
+   * @param traversal the traversal that will be used
+   * @param useNewton3 whether newton 3 optimization should be used
+   */
+  template <class ParticleFunctor, class Traversal>
+  void iteratePairwiseSoACuda(ParticleFunctor *f, Traversal *traversal, bool useNewton3 = false) {
```
Comment (SteffenSeckler) [inline src/autopas/containers/directSum/DirectSum.h]: this function looks as if it is not needed.
=> user iteratePairwise instead

Comment (SteffenSeckler) [reply src/autopas/containers/directSum/DirectSum.h]: also remove iteratePairwiseSoA

Comment (SteffenSeckler) [reply src/autopas/containers/directSum/DirectSum.h]: => you will need to adapt AutoTuner<Particle, ParticleCell>::iteratePairwiseTemplateHelper(PairwiseFunctor *f) to use iteratePairwise everywhere

File: src/autopas/containers/directSum/DirectSum.h [hunk 5]
```diff
@@ -107,6 +109,77 @@ class DirectSum : public ParticleContainer<Particle, ParticleCell> {
...
+   * Function to iterate over all pairs of particles
+   * @tparam ParticleFunctor
+   * @tparam Traversal
+   * @param f functor that describes the pair-potential
+   * @param traversal the traversal that will be used
+   */
+  template <class ParticleFunctor, class Traversal>
+  void iteratePairwise(ParticleFunctor *f, Traversal *traversal) {
+    AutoPasLog(debug, "Using traversal {} with SoA ", traversal->getTraversalType());
+    if (traversal->requiredDataLayout() == DataLayoutOption::soa) {
```
Comment (SteffenSeckler) [inline src/autopas/containers/directSum/DirectSum.h]: move this part to 
```C++
DataLayoutConverter::loadDataLayout(functor);
```
here call
```C++
traversal->initTraversal(...);
```
instead, which then calls the `DataLayoutConverter`'s `loadDataLayout` that is saved within the traversal

File: src/autopas/containers/directSum/DirectSum.h [hunk 6]
```diff
@@ -107,6 +109,77 @@ class DirectSum : public ParticleContainer<Particle, ParticleCell> {
...
+    }
+    if (auto *traversalInterface = dynamic_cast<DirectSumTraversalInterface<ParticleCell> *>(traversal)) {
+      traversalInterface->traverseCellPairs(this->_cells);
+
+    } else {
+      autopas::utils::ExceptionHandler::exception(
+          "trying to use a traversal of wrong type in DirectSum::iteratePairwise");
+    }
+
+    if (traversal->requiredDataLayout() == DataLayoutOption::soa) {
```
Comment (SteffenSeckler) [inline src/autopas/containers/directSum/DirectSum.h]: move this part to 
```C++
DataLayoutConverter::storeDataLayout(..., functor);
```
here call
```C++
traversal->endTraversal(...);
```
instead, which then calls the `DataLayoutConverter`'s `storeDataLayout` that is saved within the traversal

File: src/autopas/containers/linkedCells/traversals/C01CudaTraversal.h [hunk 7]
```diff
@@ -0,0 +1,208 @@
...
+ *
+ * The traversal calculates all cells in parallel
+ *
+ * @tparam ParticleCell the type of cells
+ * @tparam PairwiseFunctor The functor that defines the interaction of two particles.
+ * @tparam DataLayout
+ * @tparam useNewton3
+ */
+template <class ParticleCell, class PairwiseFunctor, DataLayoutOption DataLayout, bool useNewton3>
+class C01CudaTraversal : public CellPairTraversal<ParticleCell>, public LinkedCellTraversalInterface<ParticleCell> {
```
Comment (SteffenSeckler) [inline src/autopas/containers/linkedCells/traversals/C01CudaTraversal.h]: what is the reason, that this traversal is not using the C01BasedTraversal, even though you modified it a bit to fit cuda needs?

File: src/autopas/containers/linkedCells/traversals/C01CudaTraversal.h [hunk 8]
```diff
@@ -0,0 +1,208 @@
...
+    for (unsigned long y = 1; y < end_y; ++y) {
+      for (unsigned long x = 1; x < end_x; ++x) {
+        nonHaloCells.push_back(utils::ThreeDimensionalMapping::threeToOneD(x, y, z, this->_cellsPerDimension));
+      }
+    }
+  }
+#if defined(AUTOPAS_CUDA)
+  _nonHaloCells.copyHostToDevice(nonHaloCells.size(), nonHaloCells.data());
+#endif
+}  // namespace autopas
```
Comment (SteffenSeckler) [inline src/autopas/containers/linkedCells/traversals/C01CudaTraversal.h]: that's not ending the namespace

File: src/autopas/containers/verletClusterLists/VerletClusterLists.h [hunk 9]
```diff
@@ -112,6 +112,19 @@ class VerletClusterLists : public ParticleContainer<Particle, FullParticleCell<P
...
+  /**
+   * Dummy function. (Uses AoS instead)
+   * @tparam ParticleFunctor
+   * @tparam Traversal
+   * @param f functor that describes the pair-potential
+   * @param traversal the traversal that will be used
+   * @param useNewton3 whether newton 3 optimization should be used
+   */
+  template <class ParticleFunctor, class Traversal>
+  void iteratePairwiseSoACuda(ParticleFunctor* f, Traversal* traversal, bool useNewton3 = false) {
```
Comment (SteffenSeckler) [inline src/autopas/containers/verletClusterLists/VerletClusterLists.h]: should also go into a iteratePairwise() function with all other iteratePairwise calls

File: src/autopas/containers/verletListsCellBased/verletLists/VerletLists.h [hunk 10]
```diff
@@ -121,6 +122,19 @@ class VerletLists
...
+   * Dummy function. Just throws exception
+   * @tparam ParticleFunctor
+   * @tparam Traversal
+   * @param f functor that describes the pair-potential
+   * @param traversal the traversal that will be used
+   * @param useNewton3 whether newton 3 optimization should be used
+   */
+  template <class ParticleFunctor, class Traversal>
+  void iteratePairwiseSoACuda(ParticleFunctor* f, Traversal* traversal, bool useNewton3 = false) {
+    utils::ExceptionHandler::exception("VerletList Cuda not implemented");
```
Comment (SteffenSeckler) [inline src/autopas/containers/verletListsCellBased/verletLists/VerletLists.h]: move to iteratePairwise function

File: src/autopas/containers/verletListsCellBased/verletListsCells/VerletListsCells.h [hunk 11]
```diff
@@ -121,6 +122,19 @@ class VerletListsCells
...
+  /**
+   * Dummy function. Throws exception
+   * @tparam ParticleFunctor
+   * @tparam Traversal
+   * @param f functor that describes the pair-potential
+   * @param traversal the traversal that will be used
+   * @param useNewton3 whether newton 3 optimization should be used
+   */
+  template <class ParticleFunctor, class Traversal>
+  void iteratePairwiseSoACuda(ParticleFunctor* f, Traversal* traversal, bool useNewton3 = false) {
```
Comment (SteffenSeckler) [inline src/autopas/containers/verletListsCellBased/verletListsCells/VerletListsCells.h]: move

File: src/autopas/pairwiseFunctors/Functor.h [hunk 12]
```diff
@@ -8,8 +8,12 @@
 #pragma once
 
 #include "autopas/utils/AlignedAllocator.h"
+#include "autopas/utils/CudaSoA.h"
 #include "autopas/utils/ExceptionHandler.h"
 #include "autopas/utils/SoA.h"
+#if defined(AUTOPAS_CUDA)
+#include "autopas/pairwiseFunctors/LJFunctorCuda.cuh"
```
Comment (SteffenSeckler) [inline src/autopas/pairwiseFunctors/Functor.h]: this include looks very wrong :)

File: src/autopas/pairwiseFunctors/LJFunctor.h [hunk 13]
```diff
@@ -68,7 +74,13 @@ class LJFunctor : public Functor<Particle, ParticleCell, typename Particle::SoAA
         _duplicatedCalculations{duplicatedCalculation},
         _lowCorner(lowCorner),
         _highCorner(highCorner),
-        _postProcessed{false} {
+        _postProcessed {
+    false
+  }
+#if defined(AUTOPAS_CUDA)
+  , _streams(32)
```
Comment (SteffenSeckler) [inline src/autopas/pairwiseFunctors/LJFunctor.h]: this magic constant should probably be documented

File: src/autopas/pairwiseFunctors/LJFunctorCuda.cu [hunk 14]
```diff
@@ -0,0 +1,907 @@
...
+	floatType dfx = drx * fac;
+	floatType dfy = dry * fac;
+	floatType dfz = drz * fac;
+
+	fi.x += dfx;
+	fi.y += dfy;
+	fi.z += dfz;
+
+	if (n3AdditionSafe) {
+		fj->x -= dfx;
```
Comment (SteffenSeckler) [inline src/autopas/pairwiseFunctors/LJFunctorCuda.cu]: this is only allowed to be called safely if no conflicts can occur. How is this guaranteed?

File: src/autopas/pairwiseFunctors/LJFunctorCuda.cu [hunk 15]
```diff
@@ -0,0 +1,907 @@
...
+
+	typename vec3<floatType>::Type myposition = { 0, 0, 0 };
+	typename vec3<floatType>::Type myf = { 0, 0, 0 };
+	if (tid < N) {
+		myposition.x = posX[tid];
+		myposition.y = posY[tid];
+		myposition.z = posZ[tid];
+	}
+
+	for (i = block_size, tile = 0; i < N; i += block_size, ++tile) {
```
Comment (SteffenSeckler) [inline src/autopas/pairwiseFunctors/LJFunctorCuda.cu]: comments please!

File: src/autopas/pairwiseFunctors/LJFunctorCuda.cu [hunk 16]
```diff
@@ -0,0 +1,907 @@
...
+	}
+	for (i = block_size, tile = 0; i <= M; i += block_size, ++tile) {
+		int idx = tile * block_size + threadIdx.x;
+		cell2_pos_shared[threadIdx.x] = {cell2.posX[idx], cell2.posY[idx], cell2.posZ[idx]};
+		cell2_forces_shared[threadIdx.x] = {0,0,0};
+		__syncthreads();
+
+		for (int j = 0; j < block_size; ++j) {
+			const int offset = (j + threadIdx.x) & mask;
+			myf = bodyBodyFN3<floatType, true>(myposition,
```
Comment (SteffenSeckler) [inline src/autopas/pairwiseFunctors/LJFunctorCuda.cu]: why is it save to call a newton3 body here, even though no atomics are used?

File: src/autopas/pairwiseFunctors/LJFunctorCuda.cu [hunk 17]
```diff
@@ -0,0 +1,907 @@
...
+		double* posY, double* posZ, double* forceX, double* forceY,
+		double* forceZ, cudaStream_t stream);
+
+template<typename floatType>
+void CudaWrapper::SoAFunctorNoN3PairWrapper(int N, floatType* posX,
+		floatType* posY, floatType* posZ, floatType* forceX, floatType* forceY,
+		floatType* forceZ, int M, floatType* posX2, floatType* posY2,
+		floatType* posZ2, cudaStream_t stream) {
+	switch (_num_threads) {
+	case 32:
```
Comment (SteffenSeckler) [inline src/autopas/pairwiseFunctors/LJFunctorCuda.cu]: use variadic macro magic to reduce this code and others.

File: src/autopas/selectors/AutoTuner.h [hunk 18]
```diff
@@ -283,35 +283,65 @@ bool AutoTuner<Particle, ParticleCell>::iteratePairwise(PairwiseFunctor *f) {
     case DataLayoutOption::aos: {
       if (_currentConfig->_newton3 == Newton3Option::enabled) {
         if (isTuning) {
-          iteratePairwiseTemplateHelper<PairwiseFunctor, /*SoA*/ false, /*Newton3*/ true, /*tuning*/ true>(f);
+          iteratePairwiseTemplateHelper<PairwiseFunctor, /*SoA*/ DataLayoutOption::aos, /*Newton3*/ true,
```
Comment (SteffenSeckler) [inline src/autopas/selectors/AutoTuner.h]: remove comments regarding SoA

File: src/autopas/utils/ExceptionHandler.cpp [hunk 19]
```diff
@@ -5,6 +5,7 @@
  */
 
 #include "autopas/utils/ExceptionHandler.h"
+#include "ExceptionHandler.h"
```
Comment (SteffenSeckler) [inline src/autopas/utils/ExceptionHandler.cpp]: remove one of these two lines

File: cmake/modules/cuda.cmake [hunk 20]
```diff
@@ -0,0 +1,8 @@
+
+if(ENABLE_CUDA)
+target_compile_options(autopas
+        PUBLIC
+        #architecture flags and -Xcompiler to prepend to -fopenmp
+        $<$<COMPILE_LANGUAGE:CUDA>:-gencode arch=compute_61,code=sm_61 $<$<BOOL:${OPENMP}>:-Xcompiler>>
```
Comment (FG-TUM) [inline cmake/modules/cuda.cmake]: does this stuff still make sense here?
```
$<$<BOOL:${OPENMP}>:-Xcompiler>>
```

File: tests/testAutopas/tests/autopasInterface/Newton3OnOffTest.cpp [hunk 21]
```diff
@@ -207,7 +229,7 @@ void Newton3OnOffTest::countFunctorCalls(autopas::ContainerOption containerOptio
 template <class ParticleFunctor, class Container, class Traversal>
 void Newton3OnOffTest::iterate(Container container, Traversal traversal, autopas::DataLayoutOption dataLayout,
                                autopas::Newton3Option newton3, ParticleFunctor *f) {
-  switch (dataLayout) {
+  switch (traversal->requiredDataLayout()) {
```
Comment (SteffenSeckler) [inline tests/testAutopas/tests/autopasInterface/Newton3OnOffTest.cpp]: should be reverted

File: cmake/modules/other-compileroptions.cmake [hunk 22]
```diff
@@ -4,8 +4,24 @@ if (ENABLE_FAST_MATH)
     message(WARNING "Fast-Math might cause particle loss! Only use this if you know what you are doing!")
 endif()
 
-target_compile_options(autopas
-        PUBLIC
+if(ENABLE_CUDA)
```
Comment (FG-TUM) [inline cmake/modules/other-compileroptions.cmake]: if all CUDA stuff is wrapped in generator expressions this if (and some others) are not needed.

File: cmake/modules/vectorization.cmake [hunk 23]
```diff
@@ -7,30 +7,57 @@ if (USE_VECTORIZATION)
     set(VECTOR_INSTRUCTIONS "NATIVE" CACHE STRING "Vector instruction set to use (${VECTOR_INSTRUCTIONS_OPTIONS}).")
     # let ccmake and cmake-gui offer the options
     set_property(CACHE VECTOR_INSTRUCTIONS PROPERTY STRINGS ${VECTOR_INSTRUCTIONS_OPTIONS})
-
+            
+    if(ENABLE_CUDA)
```
[the remark this answers is not in the document]
Comment (FG-TUM) [reply cmake/modules/vectorization.cmake]: see my `if` comment from above

File: cmake/modules/vectorization.cmake [hunk 24]
```diff
@@ -7,30 +7,57 @@ if (USE_VECTORIZATION)
...
+        $<$<AND:$<STREQUAL:${VECTOR_INSTRUCTIONS},KNL>,$<OR:$<CXX_COMPILER_ID:GNU>,$<CXX_COMPILER_ID:Clang>>>:-march=knl>
+        $<$<AND:$<STREQUAL:${VECTOR_INSTRUCTIONS},KNL>,$<CXX_COMPILER_ID:Intel>>:-xMIC-AVX512>
+        )
+    endif()
 
 else()
 	MESSAGE(STATUS "Vectorization disabled.")
-
+            
+    if(ENABLE_CUDA)
```
Comment (FG-TUM) [inline cmake/modules/vectorization.cmake]: see my `if` comment from above

File: examples/basic-cuda/CMakeLists.txt [hunk 25]
```diff
@@ -0,0 +1,28 @@
+file(GLOB_RECURSE MY_SRC
+        "*.cpp"
+        "*.h"
+        "*.cu"
```
Comment (FG-TUM) [inline examples/basic-cuda/CMakeLists.txt]: also add .cuh files. Otherwise clion will not work properly on them.

**Comment by SteffenSeckler** [comment]
This pull request **introduces 1 alert** when merging 41752ef027d6731af307f39c9923fa8b0820a3e5 into 6a6fb2a79974feeca00c8e16b7b45448d98963ab - [view on LGTM.com](https://lgtm.com/projects/g/AutoPas/AutoPas/rev/pr-e9a75c0bf53782787cc03a8370aac9c2c8c7c43c)

**new alerts:**

* 1 for Inconsistent definition of copy constructor and assignment \(&#39;Rule of Two&#39;\)

---

*Comment posted by [LGTM.com](https://lgtm.com)*

**Comment by SteffenSeckler** [comment]
This pull request **introduces 1 alert** when merging 18ade323a733f1a82d4d224163100915d0224a3a into c8be8eb745d3f26e1f62a9c37caa2a177afc58e4 - [view on LGTM.com](https://lgtm.com/projects/g/AutoPas/AutoPas/rev/pr-53925a36dc5a26982e7f0efb9d7f97e83e301762)

**new alerts:**

* 1 for Inconsistent definition of copy constructor and assignment \(&#39;Rule of Two&#39;\)

---

*Comment posted by [LGTM.com](https://lgtm.com)*

**Comment by SteffenSeckler** [comment]
This pull request **introduces 1 alert** when merging e0383d69e158fdd6661742ea0975a445c3c8afd5 into c8be8eb745d3f26e1f62a9c37caa2a177afc58e4 - [view on LGTM.com](https://lgtm.com/projects/g/AutoPas/AutoPas/rev/pr-436593ca0ab0a9a7721c334ba17f01522cbc4178)

**new alerts:**

* 1 for Inconsistent definition of copy constructor and assignment \(&#39;Rule of Two&#39;\)

---

*Comment posted by [LGTM.com](https://lgtm.com)*

**Comment by SteffenSeckler** [comment]
This pull request **introduces 1 alert** when merging b574e2d9eef7b4ece728b71eac57a09e55aae5dd into c8be8eb745d3f26e1f62a9c37caa2a177afc58e4 - [view on LGTM.com](https://lgtm.com/projects/g/AutoPas/AutoPas/rev/pr-1f17798b2c6c4a4307042eefcf24a8acf0394f06)

**new alerts:**

* 1 for Inconsistent definition of copy constructor and assignment \(&#39;Rule of Two&#39;\)

---

*Comment posted by [LGTM.com](https://lgtm.com)*

**Comment by SteffenSeckler** [comment]
This pull request **introduces 1 alert** when merging d0f3d5d45657b1b28575165058ce13ff6ee7080e into c8be8eb745d3f26e1f62a9c37caa2a177afc58e4 - [view on LGTM.com](https://lgtm.com/projects/g/AutoPas/AutoPas/rev/pr-b1a36cc201f6a18ebdd5c4fccce00903f9da37cb)

**new alerts:**

* 1 for Inconsistent definition of copy constructor and assignment \(&#39;Rule of Two&#39;\)

---

*Comment posted by [LGTM.com](https://lgtm.com)*

File: cmake/modules/other-compileroptions.cmake (hunk 22 above)
Comment (SteffenSeckler) [reply cmake/modules/other-compileroptions.cmake]: needed for cmake <3.11 see #184

File: src/autopas/cells/FullParticleCell.h (hunk 1 above)
Comment (SteffenSeckler) [reply src/autopas/cells/FullParticleCell.h]: @julian-spahl ifdef this!

File: src/autopas/cells/FullParticleCell.h (hunk 2 above)
Comment (SteffenSeckler) [reply src/autopas/cells/FullParticleCell.h]: @julian-spahl ifdef this!

File: src/autopas/pairwiseFunctors/Functor.h (hunk 12 above)
Comment (SteffenSeckler) [reply src/autopas/pairwiseFunctors/Functor.h]: PUSH

File: src/autopas/pairwiseFunctors/LJFunctor.h (hunk 13 above)
Comment (SteffenSeckler) [reply src/autopas/pairwiseFunctors/LJFunctor.h]: push

File: src/autopas/pairwiseFunctors/LJFunctorCuda.cu (hunk 14 above)
Comment (SteffenSeckler) [reply src/autopas/pairwiseFunctors/LJFunctorCuda.cu]: ?

File: src/autopas/pairwiseFunctors/LJFunctorCuda.cu (hunk 16 above)
Comment (SteffenSeckler) [reply src/autopas/pairwiseFunctors/LJFunctorCuda.cu]: ?

File: src/autopas/pairwiseFunctors/LJFunctorCuda.cu (hunk 17 above)
Comment (SteffenSeckler) [reply src/autopas/pairwiseFunctors/LJFunctorCuda.cu]: !

File: src/autopas/cells/FullParticleCell.h (hunk 1 above)
Comment (julian-spahl) [reply src/autopas/cells/FullParticleCell.h]: typename Particle::CudaDeviceArraysType is empty when cuda is not enabled.

**Pull request merged**